### ISSUE TITLE
Transfer your keras model to pytorch

### ISSUE THREAD

**Issue by mohammad-py** [allanzelener/YAD2K#32]
Hello @allanzelener. Appreciating you in advance for your great implementation in keras. I have got one question. In some parts of your code, you use **tf.space_to_depth** function to transfer a tesnor of size (None, 38,38,64) to (None, 19,19,256). Could you please give me some hints or a numpy code to illustrate how this function works. Because as far as I know there is not such a function in **pytorch** and I would like to make it for myself. Could you please help?

**Comment by allanzelener** [comment]
An equivalent PyTorch operation is called [PixelShuffle](http://pytorch.org/docs/nn.html?highlight=pixelshuffle#pixelshuffle). You may also want to take a look at this [existing YOLOv2 PyTorcvh](https://github.com/longcw/yolo2-pytorch) implementation.

I don't think there's a NumPy equivalent to this function, but this [white paper](https://arxiv.org/abs/1609.07009) by Shi et al. has some good diagrams for what this operation is doing and how it's motivated by the upscaling convolution operation (often erroneously called deconvolution).

**Comment by mohammad-py** [comment]
Thanks for your response! In **PixelShuffle** permutation is done. I would like to know the permutation in tensorflow happens as same as pytorch or not? By the the **PixelShuffle** pytorch function is implemented based on upsample factor but here we need a downsample factor! 

The accuracy of **YOLOv2 Pytorch** is not enough for me! I would like to make it more powerful. I think your implementation is more better than that.

Could you please tell me what is the range of your network output layers value? I mean do you have negative number as the result of forward path when you forward an input image?

Another thing which I would like to know ,if it is okay for you, is that What is the order of concatenation in your implementation. I mean in concatenate_1 layer!

**Comment by mohammad-py** [comment]
Please see this [link](https://stackoverflow.com/questions/44264962/how-tf-space-to-depth-works-in-tensorflow) which i provide an example of usage this function in **stackoverflow**. In addition to reshaping, it also changes the value of the input tensors! Is it logical to change the value?

**Comment by allanzelener** [comment]
Ah, I see. I didn't realize they only implemented it for scaling factors > 1. I answered your question on Stackoverflow and opened an issue on the PyTorch Github [here](https://github.com/pytorch/pytorch/issues/1684).

None of these operations should change any values, only their position within the array.

Yes, negative outputs are theoretically possible for the width and height parameters. I haven't checked the possible ranges. It's also possible to have a bounding box with width and height such that it extends beyond the image boundary.

I'm going to close this issue now since this is more related to PyTorch than this project.
